doc/developer/platform/architecture-storage.md

```diff
@@ -107,25 +107,142 @@ All commands are durably persisted and their effect survives crashes and reboots
+The result of this collection is the primary user-facing output of the source ingestion pipeline.
+There may be other auxiliary information, like the reclocking collection, that feeds in to health checks and user-facing diagnostics.
+
+We want source ingestion to satisfy at least these three properties:
+
+1. Definiteness: this should be provided by `persist`, and isn't meant to be something we can screw up.
+2. "Fault tolerance": the source should produce specific output even in the presence of unspecified i. failures and restarts, ii. concurrency with other source implementations.
+3. "Bounded input reliance": the raw input data we have to keep around to provide these properties should be as limited as possible.
+
+## A recipe for source ingestion
+
+What follows is a recipe for source ingestion that means to provide three properties:
+
+In addition, the "recipe" aspect of this means to make is "simple" to determine that you have these properties.
+You could certainly hand roll something that also has the properties, but then you have a bunch of work to do to convince folks.
+By following the recipe, the goal is that you shouldn't have to convince anyone of anything.
+
+### Ingredients
+
+1. We have some "raw sources" of data.
+
+    These sources implement something like our `SourceReader` trait, which has a method `get_next_message()` that produces an ordered sequence of messages.
+    Messages come with some ordered "address" (e.g. a (part, offset) pair or a LSN position or byte offset).
+    The messages are emitted in an order that matches the address (the address never goes backwards).
+    The source can be initialized at an address.
+    For each address, the source produces the same message to all readers (or "NA" if it's gone).
+
+2. We have access to the `persist` framework that can record evolving collections
+
+    The framework can accept arbitrary `Append` statements and confirm when the are durably recorded.
+    It handles contended writes, picking at most one winner, and providing a single consistent history of updates.
+
+3. We have an evolving understanding of what should happen to our data from raw sources to `persist`.
+
+    We want some amount of flexibility here, as we may be surprised by the requirements of source decoding, or new opportunities.
+    The "understanding" should reflect deterministic computation.
+    Non-determinism is something we can do with an advanced recipe (more `persist` stuff, in the spirit of reclocking, where we record non-deterministic choices).
+
+### The recipe
+
+There is some prep work to do first:
+
+1.  Identify all "raw sources" that contribute to some intended output (e.g. kafka topics, metadata, schema registry).
+
+2.  "Reclock" each input, by choosing and recording in `persist` a collection of source addresses that are monotonically increasing (do not go backwards, but needn't always go forward).
```

Comment:
this one I'd guess actually is a persist shard since (I think?) there's no way for `remap` to have renditions and (I assume?) we won't need to horizontally shard it out

Reply:
I'm happy to stick with "storage collection" here. At least, I don't think there is a need to force it to be one shard, and I'm happy to let someone else determine the policy for that.

Reply:
Ah, or maybe it would be polite to have it be only one shard. I'm not sure it's needed for reclocking, but other instances of the "record non-determinism" pattern might need single-shard behavior to make sure one doesn't introduce cross-shard inconsistency (e.g. winning the append to one shard and losing to another). In any case, I'll dress up the language to distinguish things.

Reply:
much appreciated!

I'm also curious if introducing the distinction offers a chance to talk more specifically about what should happen if the shard uppers are at different timestamps. e.g. this following line later in the doc

> To start, determine `t` the `min` write frontier in `persist` for all outputs and all state.

Reply:
Yup. I've just noted that this is possible. In this case, it should all be good, in that .. in principle you should be able to restart from any time `s <= t` and your output should match any persisted output. So, while you might "fail to write" some produced updates, they should be identical to what exists (and if not, we need to start a new rendition / file some bugs). I'll put some text to that effect (that e.g. there is an `assert` one should be able to put in place about pre-committed data).

Reply:
Added to the text:
```
    Both of these changes may end up being aimed at a `persist` shard that has already committed updates at that time.
    This can be because on restart not all `persist` shards had the same frontier, or because another storage instance is contending on writes.
    The writes *should* be identical, owing to the determinism of the pipeline so far.
    If they are not identical (an instance could assert they are on any failed write) then a new rendition is required (and perhaps a bug report, if this is unexpected).
```

Reply:
I think maybe you have this text in a commit that hasn't been pushed up yet, Frank?

Reply:
I did, but intentionally. Was waiting to batch up responses.